src_c/math.c

```diff
@@ -1287,6 +1293,99 @@ vector_scale_to_length(pgVector *self, PyObject *length)
     Py_RETURN_NONE;
 }
 
+static int
+_vector_move_towards_helper(Py_ssize_t dim, double *origin_coords, double *target_coords, double distance)
+{
+    Py_ssize_t i;
+    double delta[VECTOR_MAX_SIZE];
+    double dist;
+
+    if (distance < 0) {
+        PyErr_SetString(PyExc_TypeError, 
+            "Expected distance to be greater or equal to 0");
+        return 0;
+    }
+
+    for (i = 0; i < dim; ++i)
+        delta[i] = origin_coords[i] - target_coords[i];
+
+    /* Get magnitude of Vector */
+    dist = sqrt(_scalar_product(delta, delta, dim));
+
+    if (dist <= distance || dist == 0)
+    {
+        /* Return target Vector */
+        for (i = 0; i < dim; ++i)
+            origin_coords[i] = target_coords[i];
+        return 1;
+    }
+
+    for (i = 0; i < dim; ++i)
+        origin_coords[i] = origin_coords[i] - delta[i] / dist * distance;
+
+    return 1;
+}
+
+static PyObject *
+vector_move_towards(pgVector *self, PyObject *args)
+{
+    Py_ssize_t i;
+    pgVector *target;
+    double distance;
+    pgVector *ret;
+    double self_coords[VECTOR_MAX_SIZE];
+
+    if (!PyArg_ParseTuple(args, "Od:move_towards", &target, &distance))
+        return NULL;
+    
+    if (!PySequence_AsVectorCoords(self, self_coords, self->dim))
+    {
+        PyErr_SetString(PyExc_TypeError, "Argument 1 must be a vector.");
+        return NULL;
+    }
+
+    ret = (pgVector *)pgVector_NEW(self->dim);
+    if (ret == NULL)
+        return NULL;
+
+    for (i = 0; i < self->dim; ++i)
+        ret->coords[i] = self->coords[i];
+
+    if (!_vector_move_towards_helper(self->dim, ret->coords, target->coords, distance))
+    {
+        Py_DECREF(ret);
+        return NULL;
+    }
+    return (PyObject *)ret;
+}
+
+static PyObject *
+vector_move_towards_ip(pgVector *self, PyObject *args)
+{
+    Py_ssize_t i;
+    pgVector *target;
+    double distance;
+    double self_coords[VECTOR_MAX_SIZE];
+    
+    if (!PyArg_ParseTuple(args, "Od:move_towards_ip", &target, &distance))
+        return NULL;
+    
+    if (!PySequence_AsVectorCoords(self, self_coords, self->dim))
+    {
+        PyErr_SetString(PyExc_TypeError, "Argument 1 must be a vector.");
+        return NULL;
+    }
+
+    for (i = 0; i < self->dim; ++i)
+        self->coords[i] = self->coords[i];
+
+    if (!_vector_move_towards_helper(self->dim, self->coords, target->coords, distance))
+        return NULL;
+
+    memcpy(self->coords, self->coords, self->dim * sizeof(self->coords));
```

Comment:
even this bit is not required, you cannot have same source and dest pointer for memcpy in C